### PR TITLE
enhancement: config file for persistent default flags (agentsweep.toml)

### LINKED ISSUE #121: enhancement: config file for persistent default flags (agentsweep.toml)
## Context

Every flag has to be passed on every invocation today. A team standardizing on `--no-color --format sarif --no-ignore` for CI, or a user who always wants `--source codex`, has no way to set that once.

## What to do

- Support an optional `agentsweep.toml` (or `.agentsweeprc`), read from the current directory and `~/.config/agentsweep/config.toml`, with CLI flags always winning over the file.
- Scope it to flags that are safe to default silently: `--source`, `--no-color`, `--format`, `--no-ignore`.
- Deliberately **exclude** the redaction-gating flags from config-file control: `--allow-production`, `--force`, `--no-backup` must stay explicit per invocation, so a stale config file can never silently weaken a safety gate. Enforce this in code, not just by convention.
- Document precedence (CLI flag > config file > built-in default) in README and `docs/agentsweep.1`.

## Acceptance criteria

Tests covering precedence (CLI flag overrides config file) and the excluded-flags guard (config file cannot set `--allow-production`/`--force`/`--no-backup`).

### PR DESCRIPTION
**Pull request by rishabhgit0608**
## What & why

Add an optional `agentsweep.toml` (or `.agentsweeprc`) / `~/.config/agentsweep/config.toml` so `--source`, `--no-color`, `--format`, and `--no-ignore` don't have to be repeated on every invocation. Lookup order: project file in cwd, then user config, first one found wins. CLI flags always override the config file. `--allow-production`, `--force`, and `--no-backup` can never be set from a config file — enforced in code (dropped with a warning even if present), not just by convention, so a stale or shared config file can't silently weaken a redaction safety gate.

Closes #121

## Type of change

- [ ] Bug fix
- [ ] New agent source
- [ ] New detection rule
- [x] Feature / enhancement
- [ ] Refactor / docs / chore

## Testing
$ pytest -v tests/test_config.py
16 passed in 0.42s

$ pytest -q
617 passed, 12 skipped in 6.59s

$ ruff check src/agentsweep/config.py src/agentsweep/cli.py tests/test_config.py
All checks passed!

$ mypy src/
Success: no issues found in 29 source files

$ bandit -r src/agentsweep/config.py -q
(no findings)

## Checklist

- [x] `pytest` passes locally, and CI is green on **all** platforms (Linux **and** Windows, Python 3.11–3.13)
- [x] `mypy src/` is clean — it runs on every matrix leg and is the most common reason a PR goes red
- [x] No new runtime dependency, or the PR explains why one is needed (the core install is deliberately three packages) — uses stdlib `tomllib` (Python ≥3.11, matches `requires-python`)
- [x] No real secrets, tokens, or raw history-file contents are committed — tests use synthetic, non-live examples
- [ ] Redaction / write-path changes preserve the corruption-prevention invariants (path containment, atomic replace, mandatory no-clobber backup, post-write validation, line-count preservation, audit trail) — N/A, no write-path changes
- [ ] **New detection rule?** — N/A
- [ ] **New source?** — N/A
- [x] `--json` and non-tty output stay machine-clean (no banners / styling, no `ui` import on those paths)
- [x] Did not re-introduce `force-include` in `pyproject.toml`

---

Stuck on a review comment or want to talk through an approach? [Discord](https://discord.gg/KKvtRhQvRv).

## Summary by CodeRabbit

* **New Features**
  * Added persistent TOML configuration support at project and user locations.
  * Configurable defaults include source, output format, color, and ignore behavior.
  * Command-line options override saved configuration, with predictable handling of conflicting settings.
  * Added SARIF output support through configuration, including suppression of update notices for formatted output.

* **Bug Fixes**
  * Improved validation and warnings for malformed, unsupported, incorrectly typed, or unsafe settings.

* **Documentation**
  * Documented configuration locations, lookup precedence, supported settings, restrictions, and CLI precedence.

<details><summary><h3>Greptile Summary</h3></summary>

Adds optional `agentsweep.toml` / `.agentsweeprc` / `~/.config/agentsweep/config.toml` support for persisting four flags (`source`, `no_color`, `format`, `no_ignore`). CLI flags always override the config, forbidden safety-gate keys (`allow_production`, `force`, `no_backup`) are explicitly rejected with a warning, and type mismatches are rejected rather than silently coerced.

- **`src/agentsweep/config.py`** — new loader: first-file-wins lookup, per-key type validation using `_EXPECTED_TYPES`, and defensive error handling so a broken config file never interrupts a scan.
- **`src/agentsweep/cli.py`** — post-parse config merge with guards for `--all`, `fix` verb, `--json`, and `--report`; two override escapes added (`--format human`, `--ignore`) to give users a CLI route back from config-set-only values; `_background_update_notice` extended to skip on `format="sarif"`.
- **`tests/test_config.py`** — 16 new tests covering precedence, forbidden-key rejection, type-mismatch warnings (including the secret-value-not-echoed invariant), and all merge/override paths in `_parse_run`.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge — the config file is read-only and defensive, the merge logic is well-guarded, and the safety-gate flags are enforced in code rather than by convention.

The config loader never raises, type mismatches are caught explicitly, forbidden keys are dropped with warnings, and all four merge paths in _parse_run have guards that prevent them from clobbering explicitly-set CLI flags or conflicting with incompatible verb modes. Previous review issues around --all conflicts and bool-coercion are both addressed. The test suite covers all edge cases including the secret-not-echoed invariant.

**Files Needing Attention:** No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/agentsweep/config.py | New config loader: finds project or user TOML file, validates types, drops forbidden keys with warnings; never raises. |
| src/agentsweep/cli.py | Merges config values into parsed args after argparse, with correct guards for --all, fix verb, --json, and --report; validates config-derived source/format and resolves "human" sentinel to None before downstream checks. |
| tests/test_config.py | Comprehensive tests covering load_config precedence, forbidden-key guard, type-mismatch rejection, and _parse_run merge semantics including override escapes (--format human, --ignore). |
| README.md | Documents config file feature: lookup order, TOML key mapping, precedence rules, and forbidden keys — accurate and concise. |
| docs/agentsweep.1 | Man-page entry for the three config file paths and feature description; no issues. |

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[agentsweep scan] --> B[argparse: parse CLI flags\nno_color/no_ignore default=None]
    B --> C[load_config: find first config file]
    C --> D{File found?}
    D -->|No| E[cfg equals empty dict]
    D -->|Yes| F[Parse TOML, type-check keys,\ndrop forbidden and mistyped]
    F --> E
    E --> G[Merge cfg into args\nonly where args.attr is None\nand verb guards pass]
    G --> H{args.format == human?}
    H -->|Yes| I[args.format = None]
    H -->|No| J[Validate config-derived source/format]
    I --> J
    J --> K[bool-coerce no_color and no_ignore]
    K --> L{--ignore passed?}
    L -->|Yes| M[args.no_ignore = False]
    L -->|No| N[Return args]
    M --> N
    N --> O{args.no_color is True?}
    O -->|Yes| P[ui.apply_no_color True]
    O -->|No| Q[Proceed to pipeline]
    P --> Q
```
</details>

<sub>Reviews (4): Last reviewed commit: ["feat/coder-review-fixes"](https://github.com/ishannaik/agent-sweep/commit/c2c42ad8d6cd1ca46da23fbf3073d312c5c940c9) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=49432581)</sub>

### CODE REVIEW

**Comment by Ishannaik** [comment]
Substantial piece of work and the merge-precedence logic is right. Four things before it goes in.

**Config-set SARIF gets a banner printed into it.** `_background_update_notice` skips on `args.json` (cli.py:90) but not on `args.format`. This PR makes `format = "sarif"` settable from `agentsweep.toml`, so a terminal run with a newer version on PyPI prints the `★ agentsweep N available` line to stdout ahead of the SARIF. Anything parsing that output breaks, and the user never passed a flag to explain it. Add `getattr(args, "format", None)` to that guard.

**Two config keys are one-way doors.** `--no-ignore` is `store_true` with `default=None`, and there is no `--ignore`. Set `no_ignore = true` in the file and no invocation can turn it back off, so `.agentsweepignore` is unusable until the config is edited. Same for `format`: choices are `["sarif"]` only, so once config sets it there is no CLI route back to the human report. Both need an explicit way to override from the command line, since that is the direction precedence is supposed to run.

**Warnings echo values.** `config.py:92` prints `{value!r}` on a type mismatch. On a secrets tool that is a path where a mistyped config line puts its contents on stderr every run. Print the key and the expected type, not the value.

**Tests.** Precedence, forbidden keys and type rejection are covered well. Nothing drives `main()` with config-set SARIF on a tty, which is the case that breaks. An e2e asserting stdout stays parseable would lock it down.

Separate from this PR, worth knowing: the suite is nondeterministic on your branch. Across eight runs I got two failures then six clean, and the collected count moved between 616 and 637. Same commit, same machine. Worth a look independently, since it makes CI results here hard to trust either way.

https://discord.gg/KKvtRhQvRv if you want to talk any of it through.

**Comment by rishabhgit0608** [comment]
Thanks for the thorough review — all four points are valid, fixed on the branch.

## 1. Update-notice leak into SARIF (`cli.py:90`)

Added `getattr(args, "format", None)` to the guard, right next to the existing `args.json` check. Confirmed via repro: before the fix, `format="sarif"` from config + a tty run produced `args.json=False, args.format="sarif"`, which slipped past the old guard.

## 2. One-way-door config keys

Added two explicit CLI overrides so precedence actually runs CLI > config in both directions:

- `--ignore` forces `no_ignore=False` even when a config file sets `no_ignore = true`.
- `--format human` forces `format=None` (the default report) even when a config file sets `format = "sarif"`.

Both are documented in README and the man page's `FILES` section alongside the existing precedence note.

## 3. Warning echoes the value (`config.py:92`)

The type-mismatch warning now reports only the key name and the actual/expected type names (e.g. `no_color to a str, expected a bool`), never the mistyped value itself. Added a regression test asserting a secret-looking string never appears in stderr.

## 4. Missing e2e coverage

Added `TestUpdateNoticeSkipsConfigSarif` (drives `_background_update_notice` directly with `format="sarif"` on a mocked tty + mocked PyPI response, asserts the banner doesn't print), plus regression tests for both new override flags.

## Verification

$ pytest -q
629 passed, 12 skipped in 6.57s

$ pytest -v tests/test_config.py
28 passed in 0.61s

$ ruff check src/agentsweep/cli.py src/agentsweep/config.py tests/test_config.py
All checks passed!

$ mypy src/
Success: no issues found in 29 source files

$ bandit -r src/agentsweep/config.py -q
(no findings)


I also re-ran the exact repro commands from the review by hand to confirm each fix: config-set SARIF no longer conflicts with `fix`/`--json`/`--report`/the update banner, and `--ignore`/`--format human` both restore the CLI-controllable default.

---

**On the flaky-suite note:** I haven't been able to reproduce nondeterministic collection on my machine — every run here (before and after this round of fixes) reports exactly the same count, `629 passed, 12 skipped`, with no order-dependent failures across repeated runs. I didn't touch test collection or fixture scoping in a way that would explain drift, so I'd treat it as pre-existing and independent of this PR. Happy to help dig into it separately if you can share which two tests failed or point me at a CI run that shows the collected-count swing.

**Comment by Ishannaik** [comment]
Merged.

Greptile's P1 (`source` in config silently conflicting with `--all`) does not reproduce on `c2c42ad`. Checked with `source = "codex"` in `agentsweep.toml`:

- no flags: `source` resolves to `codex`, so the config is being read.
- `--all`: `source` falls back to the parser default and `all` is `True`, so the flag wins rather than fighting the config.
- `--source cursor`: wins over the config.
- `format = "sarif"` in config with `--json`: `format` comes back `None`, so the configured format is not applied to an incompatible mode.
- The update-available banner is skipped when `format` is set, which was the SARIF-on-a-tty case from the earlier review.

Because #166 and #168 landed first, I trial-merged this into the current `main` and ran the suite on the combined result rather than on the branch alone: 664 passed, `mypy src/` clean on 29 files. The three-way interaction is fine.

CodeRabbit's docs note about the man page listing CLI spellings rather than the TOML keys (`source`, `no_color`, `format`, `no_ignore`) is worth a follow-up in `docs/agentsweep.1`. Small enough that I did not hold the merge for it.

File: src/agentsweep/cli.py [hunk 1]
```diff
@@ -367,6 +377,26 @@ def _parse_run(verb: str, rest: list[str]) -> argparse.Namespace:
...
+
+    cfg = load_config()
+    if args.source is None and "source" in cfg:
+        args.source = cfg["source"]
+    if args.no_color is None and "no_color" in cfg:
+        args.no_color = cfg["no_color"]
+    if args.format is None and "format" in cfg:
+        args.format = cfg["format"]
+    if args.no_ignore is None and "no_ignore" in cfg:
+        args.no_ignore = cfg["no_ignore"]
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply src/agentsweep/cli.py:406]: Does not reproduce. With `source = "codex"` in the config, `--all` yields `all=True` and `source` back at the parser default, so the flag wins. `--source cursor` also overrides the config.

File: src/agentsweep/config.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    forbidden_present = sorted(FORBIDDEN_KEYS & data.keys())
+    if forbidden_present:
+        joined = ", ".join(forbidden_present)
+        print(
+            f"agentsweep: warning: {path} sets {joined}, which a config file "
+            "can never control — pass these on the command line instead.",
+            file=sys.stderr,
+        )
+
+    return {key: data[key] for key in ALLOWED_KEYS if key in data}
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply src/agentsweep/config.py]: Non-bool values are rejected rather than coerced. `no_color = "yes"` prints `warning: ... sets no_color to a str, expected a bool; ignoring this key` and `load_config()` returns `{}`. The warning names the key and type only, not the value.

File: docs/agentsweep.1 [hunk 3]
```diff
@@ -170,6 +170,15 @@ if set).
 .TP
 .I ~/.agentsweep/audit.jsonl
 Append\-only audit log of every write agentsweep performs.
+.TP
+.I ./agentsweep.toml ,\ ./.agentsweeprc ,\ ~/.config/agentsweep/config.toml
+Optional config file for persistent default flags. The first one found
+(in that order) wins; values are never merged across files. Only
+.BR --source ", " --no-color ", " --format ", and " --no-ignore
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply docs/agentsweep.1:177]: Agreed, the man page lists CLI spellings where the loader reads `source`, `no_color`, `format`, `no_ignore`. Small enough that I did not hold the merge; worth a docs follow-up.

File: README.md [hunk 4]
```diff
@@ -446,6 +446,34 @@ Copy-paste path: run a scan, copy the fingerprint printed beside the false
...
+```
+
+Lookup order (first file found wins, values are never merged across files):
+
+1. `./agentsweep.toml`
+2. `./.agentsweeprc`
+3. `~/.config/agentsweep/config.toml`
+
+Precedence: **CLI flag > config file > built-in default.** Any flag you pass
+on the command line always overrides the config file.
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply README.md]: Precedence matches on this head: explicit `--source` and `--all` both override the configured value. Verified against `_parse_run` rather than the docs.

File: src/agentsweep/cli.py [hunk 5]
```diff
@@ -367,6 +377,26 @@ def _parse_run(verb: str, rest: list[str]) -> argparse.Namespace:
...
 
+    from .config import load_config
+
+    cfg = load_config()
+    if args.source is None and "source" in cfg and not args.all:
+        args.source = cfg["source"]
+    if args.no_color is None and "no_color" in cfg:
+        args.no_color = cfg["no_color"]
+    if args.format is None and "format" in cfg:
+        args.format = cfg["format"]
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply src/agentsweep/cli.py]: Handled. With `format = "sarif"` in config plus `--json`, `args.format` comes back `None`, so the configured format is not applied to the incompatible mode.
